crypto/merklearray/merkle.go

```diff
@@ -0,0 +1,133 @@
+package merklearray
+
+import (
+	"fmt"
+
+	"github.com/algorand/go-algorand/crypto"
+)
+
+type Tree struct {
+	// Level 0 is the leaves.
+	levels []layer
+}
+
+func (tree *Tree) topLayer() layer {
+	return tree.levels[len(tree.levels)-1]
+}
+
+func Build(array Array) (*Tree, error) {
+	tree := &Tree{}
+
+	var leaves layer
+	arraylen := array.Length()
+	for i := uint64(0); i < arraylen; i++ {
+		data, err := array.Get(i)
+		if err != nil {
+			return nil, err
+		}
+
+		leaves = append(leaves, crypto.HashObj(data))
+	}
+
+	if arraylen > 0 {
+		tree.levels = []layer{leaves}
+
+		for len(tree.topLayer()) > 1 {
+			tree.levels = append(tree.levels, tree.topLayer().up())
+		}
+	}
+
+	return tree, nil
+}
+
+func (tree *Tree) Root() crypto.Digest {
+	// Special case: commitment to zero-length array
+	if len(tree.levels) == 0 {
+		var zero crypto.Digest
+		return zero
+	}
+
+	return tree.topLayer()[0]
+}
+
+const validateProof = false
+
+func (tree *Tree) Prove(idxs []uint64) ([]crypto.Digest, error) {
+	if len(idxs) == 0 {
+		return nil, nil
+	}
+
+	// Special case: commitment to zero-length array
+	if len(tree.levels) == 0 {
+		return nil, fmt.Errorf("proving in zero-length commitment")
+	}
+
+	pl := make(partialLayer)
+	for _, pos := range idxs {
+		if pos >= uint64(len(tree.levels[0])) {
+			return nil, fmt.Errorf("pos %d larger than leaf count %d", pos, len(tree.levels[0]))
+		}
+
+		pl[pos] = tree.levels[0][pos]
+	}
+
+	s := &siblings{
+		tree: tree,
+	}
+
+	for l := uint64(0); l < uint64(len(tree.levels)-1); l++ {
+		var err error
+		pl, err = pl.up(s, l, validateProof)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Confirm that we got the same root hash
+	if len(pl) != 1 {
+		return nil, fmt.Errorf("internal error: partial layer produced %d hashes", len(pl))
+	}
+
+	if validateProof {
+		computedroot, ok := pl[0]
+		if !ok || computedroot != tree.topLayer()[0] {
+			return nil, fmt.Errorf("internal error: root mismatch during proof")
+		}
+	}
+
+	return s.hints, nil
+}
+
+func Verify(root crypto.Digest, elems map[uint64]crypto.Hashable, proof []crypto.Digest) error {
+	if len(elems) == 0 {
+		if len(proof) != 0 {
+			return fmt.Errorf("non-empty proof for empty set of elements")
+		}
+
+		return nil
+	}
+
+	pl := make(partialLayer)
+	for pos, elem := range elems {
+		pl[pos] = crypto.HashObj(elem)
+	}
+
+	s := &siblings{
+		hints: proof,
+	}
+
+	for l := uint64(0); len(s.hints) > 0 || len(pl) > 1; l++ {
```

Comment:
Is there an easy way to validate that `s.hints` is not needlessly long? It would be cool to prevent a DoS due to a very long proof.

Reply:
Also, suppose `elems` contained every leaf. Now we will never even look at `proof`. Is it OK that we have the property that proofs are not canonical?

Reply:
Worst case it should be log(N)*len(pl), where N is the total number of leaves in the tree.

But I think the DoS you're thinking about is probably in terms of receiving the hints in the first place?  In my code (not part of this PR) that sends a proof over the wire, I added a `allocbound=` for `msgp` on the `[]crypto.Digest` proof.

Reply:
If I send you `n` `hints`, then you will do at least `n` hashes because you will keep calling `top` in this loop, right? Worst case for a valid proof would be `log(N)*len(pl)`, but I could make a needlessly long proof. But if the `allocbound` is small enough then I guess that is not a problem.

I was just wondering if there was some easy way to be like "oh, we've already reached the root, if there are any hints left then we can stop anyway"

Reply:
Ah, I see what you mean.  Yeah, it seems like we'll do at least 1 hash per element in `s.hints`, or perhaps more (if elements are close enough to each other).

Regarding canonical proofs: I was planning to check for `len(s.hints)==0` at the end, but forgot..  Thanks for catching that!  I think with that check, proofs should be canonical.

I believe the root cause for these complications is that we don't know the size of the tree.  If we grow the root commitment to include not just the `crypto.Digest` but also a `uint64` height, then this would let us compute the bound on `len(hints)`, and as a result also bound how many hashes are computed.

Perhaps a reasonable trade-off would be to define a function to compute the bound on `len(hints)` if the caller has a bound on the number of leaves?

Reply:
Edit to the above comment: now I remember what happened to the `len(s.hints)==0` check that I was planning to add..  It got folded into the loop condition.  So if you have extra hints, the loop will keep hashing up, past the (expected) root of the tree, and the resulting `computedroot` will not match, so `Verify()` will fail.  So, I think this code has proof canonicality.

Reply:
Ah, right, we have canonicality exactly because of the potential DoS vector.

Reply:
Maybe it would be OK to just cap the implied depth of the tree at 64, since we can't represent more than 2**64 things anyway?

Reply:
That could work.

In that case, the DoS vector is potentially 64 hashes per item in `elems`.  Which means the adversary would need to get the verifier to not just consider a large proof `hints`, but also try to verify a large number of elements.  So you'd still want to bound at least one of them, but this does seem potentially better than requiring the caller to bound both..